### DISCUSSION TITLE
Current impl of counter app vs future

### DISCUSSION THREAD

**Discussion by ColtonIdle** [JakeWharton/mosaic#83] [General]
Current impl looks like this vs the future api (on the bottom)
```
fun main() = runMosaic {
  var count by mutableStateOf(0)

  setContent {
    Text("The count is: $count")
  }

  for (i in 1..20) {
    delay(250)
    count = i
  }
}
```

vs 

```
fun main() = runMosaic {
  var count by remember { mutableStateOf(0) }

  Text("The count is: $count")

  LaunchedEffect(Unit) {
    for (i in 1..20) {
      delay(250)
      count = i
    }
  }
}
```

I guess my question is how does the current version even work if it's not a launched effect. Wouldn't the for loop be constantly called since runMosaic is a composable (i think)?

**Comment by JakeWharton** [comment]
`runMosaic` is not a composable! It's a DSL scope that has two things: it is a suspending function extends from `CoroutineScope`, and it also has a `setContent` function.

All the magic are in these four lines: 

https://github.com/JakeWharton/mosaic/blob/trunk/mosaic-runtime/src/main/kotlin/com/jakewharton/mosaic/mosaic.kt#LL24-L28

**Reply by ColtonIdle** [reply]
crazy how the LaunchedEffect doesn't work. i think thats what threw me off because at first glance it seems like the current api shouldn't work (i would have thought the current api would require being wrapped in a launchedEffect).
got some more reading up to do

**Reply by JakeWharton** [reply]
It's basically the inverse. The effect is essentially surrounding the composition rather than the composition hosting the effect.
